src/js/components/TruncatingLabel.jsx

```diff
@@ -0,0 +1,225 @@
+
+        const mergedStyle = {
+            overflow: 'hidden',
+            wordWrap: 'break-word',
+            ...style
+        };
+
+        if (this.renderState !== RS_STABLE) {
+            mergedStyle.opacity = 0;
+        }
+
+        return (
+            <div style={mergedStyle} className={'TruncatingLabel ' + className}>{this.innerText}</div>
+        );
+    }
+
+    //--------------------------------------
+    //  Internal Rendering Lifecycle
+    //--------------------------------------
+
+    handleProps(props: Props) {
+
+        const {children = ""} = props;
+
+        if (typeof children === "string") {
+            this.completeText = children;
+        } else if (children === null || children === false) {
+            this.completeText = ""; // Assume content has been boolean'd out
+        } else {
+            console.warn("TruncatingLabel - Label children must be string but is", typeof children, children);
+            this.completeText = "Contents must be string";
+        }
+
+        this.renderState = RS_MEASURE;
+        this.innerText = this.completeText;
+        this.loopCount = 0;
+        this.longestGood = MINLENGTH;
+        this.shortestBad = this.innerText.length;
+    }
+
+    invalidateSize() {
+        if (!this.checkSizeRequest) {
+            this.checkSizeRequest = requestAnimationFrame(() => this.checkSize());
+        }
+    }
+
+    checkSize() {
+        this.checkSizeRequest = 0;
+
+        if (this.renderState === RS_STABLE) {
+            return; // Nothing to check, no more checks to schedule
+        }
+
+        const thisElement = ReactDOM.findDOMNode(this);
+        const {
+            scrollHeight,
+            clientHeight,
+            scrollWidth,
+            clientWidth
+        } = thisElement;
+
+        const tooBig = scrollHeight > clientHeight || scrollWidth > clientWidth;
+
+        if (this.renderState === RS_MEASURE) {
+            // First measurement since mount / props changed
+
+            if (tooBig) {
+                this.renderState = RS_FLUID;
+
+                // Set initial params for binary search of length 
+                this.longestGood = MINLENGTH;
+                this.textCutoffLength = this.shortestBad = this.innerText.length;
+            } else {
+                this.renderState = RS_STABLE;
+                this.forceUpdate(); // Re-render via react so it can update the alpha
+            }
+        }
+
+        if (this.renderState === RS_FLUID) {
+            this.loopCount++;
+
+            const lastLength = this.textCutoffLength;
+
+            let keepMeasuring;
+
+            if (this.loopCount >= MAXLOOPS) {
+                // This really shouldn't happen!
+                console.error('TruncatingLabel - TOO MANY LOOPS');
+                keepMeasuring = false;
+            } else if (lastLength <= MINLENGTH) {
+                keepMeasuring = false;
+            } else if (Math.abs(this.shortestBad - this.longestGood) < 2) {
```

Comment:
Do I interpret this line correctly as "there are no more suitable lengths to try, so stop measuring"? A comment might help explain that - it took me a moment.